src/Stack/Nix.hs

```diff
@@ -0,0 +1,180 @@
+
+-- | Run commands in a nix-shell
+module Stack.Nix
+  (execWithOptionalShell
+  ,reexecWithOptionalShell
+  ,nixCmdName
+  ,StackNixException(..)
+  ) where
+
+import           Control.Applicative
+import           Control.Exception.Lifted
+import           Control.Monad
+import           Control.Monad.Catch (throwM,MonadCatch,MonadMask)
+import           Control.Monad.IO.Class (MonadIO,liftIO)
+import           Control.Monad.Logger (MonadLogger,logDebug)
+import           Control.Monad.Reader (MonadReader,asks)
+import           Control.Monad.Trans.Control (MonadBaseControl)
+import           Data.Char (toUpper)
+import           Data.List (intercalate)
+import           Data.Maybe
+import           Data.Streaming.Process (ProcessExitedUnsuccessfully(..))
+import qualified Data.Text as T
+import           Data.Typeable
+import           Data.Version (showVersion)
+import           Network.HTTP.Client.Conduit (HasHttpManager)
+import qualified Paths_stack as Meta
+import           Prelude -- Fix redundant import warnings
+import           Stack.Constants (stackProgName)
+import           Stack.Docker (StackDockerException(OnlyOnHostException), reExecArgName)
+import           Stack.Types
+import           Stack.Types.Internal
+import           System.Environment (lookupEnv,getArgs,getExecutablePath)
+import           System.Exit (exitSuccess, exitWith)
+import           System.IO (stderr,stdin,hIsTerminalDevice)
+import           System.Process.Read
+import           System.Process.Run
+import           System.Process (CreateProcess(delegate_ctlc))
+
+
+-- | If Nix is enabled, re-runs the currently running OS command in a Nix container.
+-- Otherwise, runs the inner action.
+--
+-- This takes an optional release action which should be taken IFF control is
+-- transfering away from the current process to the intra-container one.  The main use
+-- for this is releasing a lock.  After launching reexecution, the host process becomes
+-- nothing but an manager for the call into docker and thus may not hold the lock.
+reexecWithOptionalShell
+    :: M env m
+    => IO ()
+    -> m ()
+reexecWithOptionalShell =
+    execWithOptionalShell getCmdArgs
+  where
+    getCmdArgs = do
+        args <-
+            fmap
+                (("--" ++ reExecArgName ++ "=" ++ showVersion Meta.version) :)
+                (liftIO getArgs)
+        exePath <- liftIO getExecutablePath
+        return (exePath, args)
+
+-- | If Nix is enabled, re-runs the OS command returned by the second argument in a
+-- Nix container.  Otherwise, runs the inner action.
+--
+-- This takes an optional release action just like `reexecWithOptionalShell`.
+execWithOptionalShell
+    :: M env m
+    => m (FilePath,[String])
+    -> IO ()
+    -> m ()
+execWithOptionalShell getCmdArgs inner =
+  do config <- asks getConfig
+     inShell <- getInShell
+     isReExec <- asks getReExec
+     if | inShell && not isReExec ->
+            throwM OnlyOnHostException
```

Comment:
Ok, then I'm thinking it would be better to put the test back, since I'd rather have something clearly fail than exhibit unexpected behaviour.

This shouldn't hold up the merge, but a way around this might be be to have Stack check for the STACK_IN_NIX_EXTRA_ARGS environment variable at startup and implicitly insert its contents in front of the command-line arguments (using `withArgs`), instead of passing the value of $STACK_IN_NIX_EXTRA_ARGS to the in-nix-shell `stack` command explicitly.  This could get a bit hairy though so I'd rather defer tacking this unless we need to.